src/components/ActionButton/StatefulActionButton.ts

```diff
@@ -0,0 +1,162 @@
+ * The options format for a StatefulActionButton
+ */
+export interface IStatefulActionButtonOptions {
+    /**
+     * An array containing all states used by the StatefulActionButton instance.
+     * If `switchTo` is called with an unknown state, a warning will be emitted
+     * and the transition will not occur.
+     */
+    states: [IStatefulActionButtonState, ...IStatefulActionButtonState[]];
+    /**
+     * An array containing all the state transitions allowed on the StatefulActionButton instance.
+     * If `switchTo` is called with an illegal transitions , a warning will be emitted
+     * and the transition will not occur.
+     */
+    allowedTransitions?: IStatefulActionButtonTransition[];
+    /**
+     * The initial state to be used by the StatefulActionButton instance.
+     * If nullish or missing from the states array, a warning will be emitted,
+     * no ActionButton will be constructed and the element will be hidden
+     */
+    initalState: IStatefulActionButtonState;
+}
+
+/**
+ * An action button able to handle multiple states and their transitions.
+ */
+export class StatefulActionButton {
+    static ID = 'StatefulActionButton';
+    private currentState: IStatefulActionButtonState;
+    private innerActionButton: ActionButton;
+
+    constructor(public element: HTMLElement, public options: IStatefulActionButtonOptions, public bindings?: IResultsComponentBindings) {
+        if (!this.areOptionsValid()) {
+            return;
+        }
+        this.currentState = this.options.initalState;
+        this.currentState.onStateEntry?.apply(this);
+        this.innerActionButton = new ActionButton(element, { ...this.options.initalState, click: this.handleClick.bind(this) }, bindings);
+    }
+
+    /**
+     * Switch the state of the instance if the state and the transition between the current and new state are allowed.
+     * @param state a state to try to switch to
+     */
+    public switchTo(state: IStatefulActionButtonState) {
+        if (!this.isTransitionAllowed(state)) {
+            console.warn(
+                `State '${state.name}' is not allowed on this StatefulActionButton.\nEnsure to use the object references used at the instantiation`
+            );
+            return;
+        }
+        this.currentState.onStateExit?.apply(this);
+        state.onStateEntry?.apply(this);
+        this.innerActionButton.updateIcon(state.icon);
+        this.innerActionButton.updateTooltip(state.tooltip);
+        this.currentState = state;
+    }
+
+    /**
+     * Return the current state of the instance.
+     */
+    public getCurrentState() {
+        return this.currentState;
+    }
+
+    /**
+     * Check if the options given to the constructor are valid.
+     * If not, it will also display the appropriate warning.
+     */
+    private areOptionsValid(): boolean {
+        if (!this.options.states?.length) {
+            console.warn('The stateful action button cannot render if no states are defined.');
```

Comment:
Those 3 strings should be help in a const or a static no ?

Reply:
We don't tend to put error message in const, and I don't think there's positive value to that.